src/bin/pack.rs

```diff
@@ -0,0 +1,362 @@
+                    .map(|res| res.map(|e| e.path()))
+                    .collect::<Result<Vec<_>, Error>>()?;
+                children.sort_unstable_by(|a, b| a.file_name().cmp(&b.file_name()));
+
+                let mut entries = BTreeMap::new();
+
+                for child in &children {
+                    let child_name = child.file_name().unwrap().to_str().unwrap();
+                    if config.ignored_file(child_name) {
+                        warn!("skipping ignored file {:?}", child_name);
+                        continue;
+                    }
+                    let name: String;
+                    match xattr::get(&child, "user.original_name") {
+                        Ok(Some(original_name)) if config.allow_xattr => {
+                            let old_name = str::from_utf8(&original_name).unwrap();
+                            if !config.valid_name(old_name) {
+                                // original name must have been munged, so restore original
+                                name = old_name.to_string();
+                            } else {
+                                // original name wasn't munged, keep the current name
+                                // in case it was renamed
+                                name = child_name.to_string();
+                            }
+                        }
+                        Ok(_) | Err(_) => {
+                            // use current name because either --no-xattr is set,
+                            // xattr is None, or getting xattr on file (like ._ files) errors
+                            name = child_name.to_string();
+                        }
+                    }
+                    self.depth += 1;
+                    let value = self.pack(child.clone(), &config)?;
+                    self.depth -= 1;
+                    if let Some(value) = value {
+                        entries.insert(name, value);
+                    }
+                }
+
+                Ok(Some(V::from_named_dir(entries, &config)))
+            }
+            "list" => {
+                let mut children = fs::read_dir(path.clone())?
+                    .map(|res| res.map(|e| e.path()))
+                    .collect::<Result<Vec<_>, Error>>()?;
+                children.sort_unstable_by(|a, b| {
+                    a.file_name()
+                        .unwrap()
+                        .to_str()
+                        .unwrap()
+                        .cmp(&b.file_name().unwrap().to_str().unwrap())
+                });
+
+                let mut entries = Vec::with_capacity(children.len());
+
+                for child in children {
+                    let child_name = child.file_name().unwrap().to_str().unwrap();
+                    if config.ignored_file(child_name) {
+                        warn!("skipping ignored file {:?}", child_name);
+                        continue;
+                    }
+                    self.depth += 1;
```

Comment:
Could `depth` be a parameter to `::pack()`? This kind of mutation is risky (e.g., if we were to process subtrees concurrently).